File: 코드샘플/메타토이드래곤즈/코코스크리에이터/Scripts/Data/StageData.ts

import { _decorator } from 'cc';
import { DataBase } from './DataBase';

export class StageBaseData extends DataBase
{
    protected world: number = -1;
    public get WORLD(): number { return this.world }
    public set WORLD(value: number) { this.world = value }

    protected type: number = -1;
    public get TYPE(): number { return this.type }
    public set TYPE(value: number) { this.type = value }

    protected difficult: number = -1;
    public get DIFFICULT(): number { return this.difficult }
    public set DIFFICULT(value: number) { this.difficult = value }

    protected stage: number = -1;
    public get STAGE(): number { return this.stage }
    public set STAGE(value: number) { this.stage = value }

    protected name: number = -1;
    public get NAME(): number { return this.name }
    public set NAME(value: number) { this.name = value }

    protected image: string = "";
    public get IMAGE(): string { return this.image }
    public set IMAGE(value: string) { this.image = value }

    protected costtype: string = "";
    public get COST_TYPE(): string { return this.costtype }
    public set COST_TYPE(value: string) { this.costtype = value }

    protected costvalue: number = -1;
    public get COST_VALUE(): number { return this.costvalue }
    public set COST_VALUE(value: number) { this.costvalue = value }

    protected clearcount: number = -1;
    public get CLEAR_COUNT(): number { return this.clearcount }
    public set CLEAR_COUNT(value: number) { this.clearcount = value }

    protected time: number = -1;
    public get TIME(): number { return this.time }
    public set TIME(value: number) { this.time = value }

    protected spawn: number = -1;
    public get SPAWN(): number { return this.spawn }
    public set SPAWN(value: number) { this.spawn = value }

    protected accountexp: number = -1;
    public get ACCOUNT_EXP(): number { return this.accountexp }
    public set ACCOUNT_EXP(value: number) { this.accountexp = value }

    protected charexp: number = -1;
    public get CHAR_EXP(): number { return this.charexp }
    public set CHAR_EXP(value: number) { this.charexp = value }

    protected rewardgold: number = -1;
    public get REWARD_GOLD(): number { return this.rewardgold }
    public set REWARD_GOLD(value: number) { this.rewardgold = value }

    protected rewarditem: number = -1;
    public get REWARD_ITEM(): number { return this.rewarditem }
    public set REWARD_ITEM(value: number) { this.rewarditem = value }

    protected rewarditemcount: number = -1;
    public get REWARD_ITEM_COUNT(): number { return this.rewarditemcount }
    public set REWARD_ITEM_COUNT(value: number) { this.rewarditemcount = value }

    protected rewardstar: number[] = Array<number>(3);
    public get REWARD_STAR(): number[] { return this.rewardstar }
    public set REWARD_STAR(value: number[]) { this.rewardstar = value }

    protected chargecosttype: string = "";
    public get CHARGE_COST_TYPE(): string { return this.chargecosttype }
    public set CHARGE_COST_TYPE(value: string) { this.chargecosttype = value }

    protected chargecostvalue: number = -1;
    public get CHARGE_COST_VALUE(): number { return this.chargecostvalue }
    public set CHARGE_COST_VALUE(value: number) { this.chargecostvalue = value }

    protected chargecount: number = -1;
    public get CHARGE_COUNT(): number { return this.chargecount }
    public set CHARGE_COUNT(value: number) { this.chargecount = value }

    protected chargemax: number = -1;
    public get CHARGE_MAX(): number { return this.chargemax }
    public set CHARGE_MAX(value: number) { this.chargemax = value }

    protected unlockmission: number = -1;
    public get UNLOCK_MISSION(): number { return this.unlockmission }
    public set UNLOCK_MISSION(value: number) { this.unlockmission = value }
}

File: 코드샘플/메타토이드래곤즈/유니티/System/Data/Datas/StageData.cs
using Newtonsoft.Json.Linq;
using System.Collections.Generic;

namespace SandboxNetwork
{
    public class StageBaseData : TableData<DBStage_base>
    {        
        static private StageTable table = null;
        static public StageBaseData Get(string key)
        {
            if (table == null)
                table = TableManager.GetTable<StageTable>();

            return table.Get(key);
        }

        static public StageBaseData Get(int key)
        {
            return Get(key.ToString());
        }

        static public List<StageBaseData> GetByAdventureWorld(int world)
        {
            return GetByWorld(world, (StageDifficult)CacheUserData.GetInt("adventure_difficult", 1));
        }

        static public List<StageBaseData> GetByWorld(int world, StageDifficult difficult = StageDifficult.NORMAL)
        {
            if (table == null)
                table = TableManager.GetTable<StageTable>();

            return table.GetByWorld(world, difficult);
        }

        static public StageBaseData GetByAdventureWorldStage(int world, int stage)
        {
            return GetByWorldStage(world, stage, (StageDifficult)CacheUserData.GetInt("adventure_difficult", 1));
        }

        static public StageBaseData GetByWorldStage(int world, int stage, StageDifficult difficult = StageDifficult.NORMAL)
        {
            if (table == null)
                table = TableManager.GetTable<StageTable>();

            return table.GetByWorldStage(world, stage, difficult);
        }

        static public List<StageBaseData> GetDailyStage(int day, int stage)
        {
            if (table == null)
                table = TableManager.GetTable<StageTable>();

            List<StageBaseData> ret = new List<StageBaseData>();
            foreach (var data in table.GetDailyStage())
            {
                if (data.WORLD == day && data.STAGE == stage)
                    ret.Add(data);
            }

            return ret;
        }

        public int KEY => Int(Data.UNIQUE_KEY);
        public eStageType TYPE => (eStageType)Data.TYPE;
        public StageDifficult DIFFICULT => (StageDifficult)Data.DIFFICULT;
        public int WORLD => Data.WORLD;
        public int STAGE => Data.STAGE;
        public int _NAME => Data._NAME;
        public string IMAGE => Data.IMAGE;
        public int PROPERTY_PAGE => Data.PROPERTY_PAGE;
        public string COST_TYPE => Data.COST_TYPE;
        public int COST_VALUE => Data.COST_VALUE;
        public int CLEAR_COUNT => Data.CLEAR_COUNT;
        public int TIME => Data.TIME;
        public int SPAWN => Data.SPAWN;
        public int REWARD_ACCOUNT_EXP => Data.REWARD_ACCOUNT_EXP;
        public int REWARD_CHAR_EXP => (int)(Data.REWARD_CHAR_EXP * ServerOptionData.GetFloat("dragon_exp", 1.0f));
        public int REWARD_GOLD => Data.REWARD_GOLD;
        public List<Asset> REWARD_ITEMS
        {
            get
            {
                List<Asset> ret = new List<Asset>();
                float rate = 1.0f;
                switch (TYPE)
                {
                    case eStageType.ADVENTURE:
                        rate = ServerOptionData.GetFloat("adventure_reward", 1.0f);
                        break;
                    case eStageType.DAILY_DUNGEON:
                        rate = ServerOptionData.GetFloat("daily_reward", 1.0f);
                        break;
                    case eStageType.WORLD_BOSS:
                        rate = ServerOptionData.GetFloat("raid_reward", 1.0f);
                        break;
                }

                var groupData = ItemGroupData.Get(Data.REWARD_ITEM);
                if (groupData != null)
                {
                    foreach (var item in groupData)
                    {
                        if (item == null)
                            continue;

                        if (TYPE == eStageType.DAILY_DUNGEON)
                        {
                            var baseData = ItemBaseData.Get(item.Reward.ItemNo);
                            if (baseData != null && baseData.KIND != eItemKind.SKILL_UP)
                            {
                                ret.Add(new Asset(item.Reward.GoodType, item.Reward.ItemNo, (int)(item.Reward.Amount)));
                                continue;
                            }
                        }

                        ret.Add(new Asset(item.Reward.GoodType, item.Reward.ItemNo, (int)(item.Reward.Amount * rate)));
                    }
                }

                return ret;
            }
        }

        public int HOLDER_REWARD_ITEM => Data.HOLDER_REWARD_ITEM;
        public int REWARD_ITEM_COUNT => Data.REWARD_ITEM_COUNT;

        private List<int> rewards = null;
        public List<int> REWARDS
        {
            get
            {
                if(rewards == null)
                {
                    rewards = new List<int>();
                    rewards.Add(Data.FIRST_REWARD_STAR_1);
                    rewards.Add(Data.FIRST_REWARD_STAR_2);
                    rewards.Add(Data.FIRST_REWARD_STAR_3);
                }
                return rewards;
            }
        }
        public string CHARGE_COST_TYPE => Data.CHARGE_COST_TYPE;
        public int CHARGE_COST_VALUE => Data.CHARGE_COST_VALUE;
        public int CHARGE_COUNT => Data.CHARGE_COUNT;
        public int CHARGE_MAX => Data.CHARGE_MAX;
        public int UNLOCK_MISSION => Data.UNLOCK_MISSION;

        public List<ItemGroupData> GetStarRewards(int star)
        {
            if(star < GetStarRewardCount())
                return ItemGroupData.Get(REWARDS[star]);
            
            return new List<ItemGroupData>();
        }

        public int GetStarRewardCount()
        {
            return REWARDS.Count;
        }
    }
}